plugins/router/f5/f5.go

```diff
@@ -0,0 +1,1743 @@
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("Datagroup %s created.", datagroupName)
+
+	return nil
+}
+
+func (f5 *f5LTM) ensureIRuleExists(iRuleName, iRule string) error {
+	glog.V(4).Infof("Checking whether iRule %s exists...", iRuleName)
+
+	iRuleUrl := fmt.Sprintf("https://%s/mgmt/tm/ltm/rule/%s", f5.host, iRuleName)
+
+	err := f5.get(iRuleUrl, nil)
+	if err != nil && err.(F5Error).httpStatusCode != 404 {
+		// 404 is expected, but anything else really is an error.
+		return err
+	}
+
+	if err == nil {
+		glog.V(4).Infof("iRule %s already exists; nothing to do.", iRuleName)
+		return nil
+	}
+
+	glog.V(4).Infof("IRule %s does not exist; creating it now...", iRuleName)
+
+	iRulesUrl := fmt.Sprintf("https://%s/mgmt/tm/ltm/rule", f5.host)
+
+	iRulePayload := struct {
+		Name string `json:"name"`
+		Code string `json:"apiAnonymous"`
+	}{
+		Name: iRuleName,
+		Code: iRule,
+	}
+
+	err = f5.post(iRulesUrl, iRulePayload, nil)
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("IRule %s created.", iRuleName)
+
+	return nil
+}
+
+func (f5 *f5LTM) ensureVserverHasIRule(vserverName, iRuleName string) error {
+	glog.V(4).Infof("Checking whether vserver %s has iRule %s...",
+		vserverName, iRuleName)
+
+	vserverUrl := fmt.Sprintf("https://%s/mgmt/tm/ltm/virtual/%s",
+		f5.host, vserverName)
+
+	res := struct {
+		Rules []string `json:"rules"`
+	}{}
+
+	err := f5.get(vserverUrl, &res)
+	if err != nil {
+		return err
+	}
+
+	commonIRuleName := fmt.Sprintf("/Common/%s", iRuleName)
+
+	for _, name := range res.Rules {
+		if name == commonIRuleName {
+			glog.V(4).Infof("Vserver %s has iRule %s associated with it;"+
+				" nothing to do.",
+				vserverName, iRuleName)
+			return nil
+		}
+	}
+
+	glog.V(4).Infof("Adding iRule %s to vserver %s...", iRuleName, vserverName)
+
+	vserverRulesPayload := struct {
+		Rules []string `json:"rules"`
+	}{
+		Rules: []string{sslPassthroughIRuleName},
+	}
+
+	err = f5.patch(vserverUrl, vserverRulesPayload, nil)
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("IRule %s added to vserver %s.", iRuleName, vserverName)
+
+	return nil
+}
+
+// Initialize ensures that OpenShift-specific configuration is in place
+// on the F5 BIG-IP host.  In particular, this method creates a traffic
+// policy and associates it with the vservers if necessary.
+func (f5 *f5LTM) Initialize() error {
+	err := f5.ensurePolicyExists(httpPolicyName)
+	if err != nil {
+		return err
+	}
+
+	if f5.httpVserver != "" {
+		err = f5.ensureVserverHasPolicy(f5.httpVserver, httpPolicyName)
+		if err != nil {
+			return err
+		}
+	}
+
+	if f5.httpsVserver != "" {
+		err = f5.ensurePolicyExists(httpsPolicyName)
+		if err != nil {
+			return err
+		}
+
+		err = f5.ensureVserverHasPolicy(f5.httpsVserver, httpsPolicyName)
+		if err != nil {
+			return err
+		}
+
+		if f5.httpVserver != "" {
+			err = f5.ensureVserverHasPolicy(f5.httpVserver, httpPolicyName)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
Wow.  Yeah, that's weird.  My next fixup commit will re-arrange the logic to delete the redundant `f5.ensureVserverHasPolicy(f5.httpVserver, httpPolicyName)`.

I'll also move `f5.ensurePolicyExists(httpsPolicyName)`, `f5.ensureDatagroupExists(passthroughRoutesDataGroupName)`, `f5.ensureDatagroupExists(passthroughHostsDataGroupName)`, and `f5.ensureIRuleExists(sslPassthroughIRuleName, sslPassthroughIRule)` outside the `if f5.httpsVserver != ""` block because `addRoute` and other functions assume that the policy and datagroups exist and will act and generate a lot of errors if they do not, so these things needs to exist even if they are not associated with a vserver.